opta/linker.py

```diff
@@ -0,0 +1,416 @@
+from collections.abc import MutableSequence
+from dataclasses import dataclass, field
+from typing import Any, Dict, FrozenSet, Iterable, List, Set
+
+from opta.layer2 import Layer
+from opta.link import Link
+from opta.link_spec import InputLinkSpec, OutputLinkSpec
+from opta.module2 import Module
+from opta.module_spec import ModuleSpec
+from opta.utils.ref import (
+    InterpolatedReference,
+    Reference,
+    SimpleInterpolatedReference,
+    get_all_references,
+    parse_ref_string,
+)
+from opta.utils.visit import Visitor, fill_missing_list_or_dict
+
+_ModuleMap = Dict[str, Module]
+
+
+@dataclass
+class LinkResult:
+    execution_order: List[FrozenSet[str]] = field(default_factory=list)
+    interpolations: Dict[str, Dict[Reference, InterpolatedReference]] = field(
+        default_factory=dict
+    )
+
+
+class Linker:
+    def __init__(self, module_specs: Iterable[ModuleSpec]):
+        self._module_specs = {spec.name: spec for spec in module_specs}
+
+    def process(self, layer: Layer) -> LinkResult:
+        result = LinkResult()
+        module_map = self._module_map(layer.modules)
+
+        result.interpolations = self._pass_interpolation(module_map)
+
+        # Pass: TODO Run validation of input, but ignore defaults or required; assume refs are valid
+        # Pass: TODO Make sure that there are no interpolations in the link params
+
+        # Pass: TODO Run validation of input, but ignore defaults or required; assume refs are valid
+        # Pass: TODO Make sure that there are no interpolations in the link params
```

Comment:
It's a bit overwhelming the amount of TODO in this PR :) 

Do you think we can introduce some prioritization?

You can define some high level milestones in docs/module-api.md
ex:
- milestone 1: get aws-base working creation
- milestone 2: get aws-base deletion
- milestone 3: get aws-base interpolation
- ... X: deprecate old module api

And then we can revisit the TODO to include the milestone where it will be needed.
- TODO M3:  Make sure that there are no interpolations in the link params

That will give some context and allow the rest of the team to start implementing these TODO
What do you think?